modules/homa-lite/src/lib.rs

```diff
@@ -783,44 +784,59 @@ pub mod module {
 
 				// Check if the redeem request exists
 				if let Some((request_amount, extra_fee)) = Self::redeem_requests(&redeemer) {
-					Self::match_mint_with_redeem_request(
+					let actual_liquid_redeemed = Self::match_mint_with_redeem_request(
 						minter,
 						&redeemer,
 						request_amount,
 						extra_fee,
-						&mut liquid_remaining,
-						&mut new_balances,
+						liquid_remaining,
 					)?;
+					liquid_remaining = liquid_remaining.saturating_sub(actual_liquid_redeemed);
+					new_balances.push((
+						redeemer,
+						request_amount.saturating_sub(actual_liquid_redeemed),
+						extra_fee,
+					));
 				}
 			}
 
 			// Update storage to the new balances. Remove Redeem requests that have been filled.
 			Self::update_redeem_requests(&new_balances);
 
-			// Redeem request storage has now been updated.
 			new_balances.clear();
-
 			let mut redeem_requests_limit_remaining = T::MaximumRedeemRequestMatchesForMint::get();
-			// Iterate all remaining redeem requests now.
-			for (redeemer, (request_amount, extra_fee)) in RedeemRequests::<T>::iter() {
-				// If all the currencies are minted, return.
-				if liquid_remaining.is_zero() || redeem_requests_limit_remaining.is_zero() {
-					break;
-				}
-				Self::match_mint_with_redeem_request(
-					minter,
-					&redeemer,
-					request_amount,
-					extra_fee,
-					&mut liquid_remaining,
-					&mut new_balances,
-				)?;
-				redeem_requests_limit_remaining -= 1;
+			if !liquid_remaining.is_zero() && !redeem_requests_limit_remaining.is_zero() {
+				// Define the function to be executed for each of the redeem request:
+				// Redeem the given requests with the current minting request.
+				//
+				// Params: redeemer: T::AccountId, request_amount: Balance, extra_fee: Permill
+				// Returns: Result<(new_redeem_amount, should_break), Error>
+				//
+				// Capturing: &mut liquid_remaining
+				//            &mut redeem_requests_limit_remaining
+				// If all the currencies are minted, return `should_break` as true.
+				let mut f = |redeemer, request_amount, extra_fee| -> Result<(bool, Balance), DispatchError> {
```

Comment:
Capturing mutable variables in closure is not readable and prone to bugs. Please refactor it to be no closure and mutable params.

Reply:
Agree. WIth the closure and captures the whole thing is a mess.
I've significantly refactored the code structure. Please have a look at the updated structure.
Al closures are removed.